packages/experiments/src/components/Coachmark/Coachmark.tsx

```diff
@@ -0,0 +1,226 @@
+          </div>
+        </div>
+      </PositioningContainer>
+    );
+  }
+
+  public componentWillReceiveProps(newProps: ICoachmarkProps) {
+    if (this.props.isCollapsed && !newProps.isCollapsed) {
+      // The coachmark is about to open
+      this._openCoachmark();
+    }
+  }
+
+  public componentDidMount() {
+    // If we have already accessed the width and height
+    // We do not wan't to do it again.
+    if (this.state.isMeasuring) {
+      this._async.setTimeout(() => {
+        if ((this.state.entityInnerHostRect.width + this.state.entityInnerHostRect.width) === 0) {
+          this.setState({
+            isMeasuring: false,
+            entityInnerHostRect: {
+              width: this._entityInnerHostElement.offsetWidth,
+              height: this._entityInnerHostElement.offsetHeight
+            }
+          });
+
+          this.forceUpdate();
+        }
+
+        // Initialize element in proximity now that initial
+        // measurements have been taken
+        this._isElementInProximity(this._entityInnerHostElement);
+      }, 10);
+    }
+  }
+
+  private _openCoachmark() {
+    this.setState({
+      isCollapsed: false
+    });
+
+    this._translateAnimationContainer.addEventListener('animationstart', () => {
+      if (this.props.onAnimationBegin) {
+        this.props.onAnimationBegin();
+      }
+    });
+
+    this._translateAnimationContainer.addEventListener('animationend', () => {
+      if (this.props.onAnimationEnd) {
+        this.props.onAnimationEnd();
+      }
+    });
+  }
+
+  private _isElementInProximity(targetElement: HTMLElement, mouseProximityOffset: number = 0) {
+    /**
+     * An array of cached ids returned when setTimeout runs during
+     * the window resize event trigger.
+     */
+    let timeoutIds: number[] = [];
+
+    /**
+     * The target element the mouse would be in
+     * proximity to
+     */
+    let targetElementRect: ClientRect;
+
+    // When the window resizes we want to async
+    // get the bounding client rectangle.
+    // Every time the event is triggered we wan't to
+    // setTimeout and then clear any previous instances
+    // of setTimeout.
+    this._events.on(window, 'resize', () => {
+      timeoutIds.forEach((value) => {
+        clearInterval(value)
+      });
+
+      timeoutIds.push(this._async.setTimeout(() => {
+        targetElementRect = targetElement.getBoundingClientRect();
+      }, 100));
+    });
+
+    // Everytime the document's mouse move is triggered
+    // we want to check if inside of an element and
+    // set the state with the result.
+    this._events.on(document, 'mousemove', (e: MouseEvent) => {
+      let mouseY = e.pageY;
+      let mouseX = e.pageX;
+
+      this.setState({
+        isMouseInProximity: this._isInsideElement(targetElementRect, mouseX, mouseY)
+      });
+    });
+  }
+
+  private _isInsideElement(elementRect: ClientRect, mouseX: number, mouseY: number, mouseProximityOffset: number = 0): boolean {
```

Comment:
I think this should also be a helper function somewhere.

Reply:
Yeah i think i will save that for a future pull request. I think this helper is super specific and we could make it a bit more generic by adding a couple methods to Rectangle and potentially one for Circle.